saleor/graphql/product/mutations/attributes.py

```diff
@@ -122,6 +130,58 @@ def check_operations_not_assigned_already(
             )
             errors["operations"].append(error)
 
+    @classmethod
+    def check_type_and_variant_coherence(cls, errors, product_attrs_data):
+        for attr_pk, type_, variant_selection in product_attrs_data:
+            if variant_selection:
+                msg = (
+                    f"Attribute with pk: '{attr_pk}' with different type "
+                    f"than 'VARIANT' (found: '{type_}') cannot be assigned with "
+                    "variant_selection: true."
+                )
+                error = ValidationError(
+                    msg,
+                    code=ProductErrorCode.ATTRIBUTE_CANNOT_BE_ASSIGNED,
+                    params={"attributes": attr_pk},
+                )
+                errors["operations"].append(error)
+
+    @classmethod
+    def check_allowed_types(cls, errors, variant_attrs_data):
+        variant_attrs_pks = [
+            pk for pk, _, variant_selection in variant_attrs_data if variant_selection
+        ]
+        qs = (
+            attribute_models.Attribute.objects.filter(
+                pk__in=variant_attrs_pks,
+            )
+            .exclude(input_type__in=AttributeInputType.ALLOWED_IN_VARIANT_SELECTION)
+            .values_list("pk", "name", "input_type")
+        )
+
+        invalid_attributes = list(qs)
+
+        if invalid_attributes:
+            msg = ", ".join(
+                [
+                    f"<Attribute name: '{name}', input_type: '{input_type}'>"
+                    for _, name, input_type in invalid_attributes
+                ]
+            )
```

Comment:
Not sure if we need to include all attributes in the message; invalid IDs are already included in error's context (`params`).

Reply:
I have removed entire join and change the message little bit. Now it will be much shorter.